### DISCUSSION TITLE
Is there a way to merge two completely different data sets

### DISCUSSION THREAD

**Discussion by pranavbhatnagar11** [moj-analytical-services/splink#830] [General]
Is there a way to merge two completely different data sets. There can be 3-4 common columns. If there is a not a direct match, then fuzzy matching can work. Just how fuzzymatcher worked. 
I looked into linking two tables of persons, but I think it shoots an error for using different data frames. Is there a way!

**Comment by RobinL** [comment]
Hello!  Prior to ingesting data into Splink, you need to standardise the data so that each input dataset contains the same column names.  

For anyone reading this in future, this is now part of the tutorial [here](https://moj-analytical-services.github.io/splink/demos/01_Prerequisites.html)

**Reply by Philistino** [reply]
Hi Robin, 

Firstly, I want to say that this is a great project and you have clearly put a lot of work into it.

I think noting the prerequisites of the data shapes and labels in the readme would be very helpful. I wasted a bit of time figuring out that the column names had to match across datasets and that a column named unique_id is required. I'm sure others will struggle with the same issue, and early and unexplained errors can put off new users.

To integrate this tool with existing workflows, it would be really helpful to give the user more flexibility in the input data for matching across datasets. A few features that I think would be super helpful:

- Allow the user to provide a column mapping between two datasets, perhaps as a dictionary. This is particularly relevant when using a persistent datastore. 
- As a user, it doesn't make sense to specify left and right column names (e.g., "l.address = r.address") in the blocking rules if they have to be named the same anyway ("address" in the prior example). Maybe I am missing something here.
- Ignore any additional columns that are not part of the settings specification so that the user doesn't have to pare down their dataframes to only the shared columns across dataframes. Perhaps an easy solution is to ignore columns not included in the 'comparisons' section of the settings.
- Allow the user to identify their own unique identifier column. If they do not provide one, you could default to the index of the dataframe/database table.

Once again, thank you for the awesome project. As someone who uses UK public data sources that lack a common unique identifier, I really appreciate what you are doing!

Best regards,
Philistino

**Reply by RobinL** [reply]
Hi Philstino,

Thanks for taking the time to write down this feedback - it's really valuable to us, so very appreciated and there are some great suggestions here.

A few notes, but we should convert some of this into issues:

> Allow the user to identify their own unique identifier column. If they do not provide one, you could default to the index of the dataframe/database table.

This one can be done with the [unique_id_column_name](https://moj-analytical-services.github.io/splink/settings_dict_guide.html#unique_id_column_name) setting

> Ignore any additional columns that are not part of the settings specification so that the user doesn't have to pare down their dataframes to

This is a great idea and not too difficult to implement.  We should do it

> As a user, it doesn't make sense to specify left and right column names

In most cases, you're right.  However, there are some cases in which the user does need to be explicit.  

For example, in the [financial transactions example](https://moj-analytical-services.github.io/splink/demos/example_transactions.html), the entity type is a money transfer between accounts.  Money is sent from one account and arrives in the other with a delay.  In this case, similarity is defined as the length of time it takes for the money to turn up and this is _not_ symmetric.  

Another example is that sometimes the use may wish to provide a more complex SQL expression here (`substr(l.name,1,2) = substr(r.name,1,2)` or a non-equality condition (`levenstein(l.name, r.name) < 2)`).  It would introduce a lot of complexity into the codebase to be able to support all of these possibilities. whilst allowing the user to provide the simpler syntax as you suggest.


> Allow the user to provide a column mapping between two datasets, perhaps as a dictionary. This is particularly relevant when using a persistent datastore.

This is a good idea.  We've rejected in the past because Splink used to be PySpark only, and at that time, it was relatively simple to do the renames.  

But I totally take your point about persistent datastores.  Also, due to how Splink 3 is implemented, it would now be relatively easy to achieve by modifying the code that creates `__splink__df_concat_with_tf`.

 I think there's a bit of a challenge to find a sensible user-facing API to do this.  The complexity arises from the fact users can potentially provide _n_ input dataframes

**Reply by Philistino** [reply]
Hi Robin,

Thank you for the thoughtful reply. I apologize for not responding sooner. I was hoping to review the repo and submit a PR to implement one or more of my suggestions above. Unfortunately, I am in a busy period and I cannot assist right now. I have this project starred and I look forward to seeing it develop!

Philistino
